sentinel-core/src/main/java/com/alibaba/csp/sentinel/slots/block/degrade/circuitbreaker/AbstractCircuitBreaker.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 1999-2019 Alibaba Group Holding Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.alibaba.csp.sentinel.slots.block.degrade.circuitbreaker;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.alibaba.csp.sentinel.slots.block.degrade.DegradeRule;
+import com.alibaba.csp.sentinel.slots.block.degrade.DegradeRuleManager;
+import com.alibaba.csp.sentinel.util.AssertUtil;
+import com.alibaba.csp.sentinel.util.TimeUtil;
+
+/**
+ * @author Eric Zhao
+ * @since 1.8.0
+ */
+public abstract class AbstractCircuitBreaker implements CircuitBreaker {
+
+    protected final DegradeRule rule;
+    protected final int retryTimeoutMs;
+
+    private final EventObserverRegistry observerRegistry;
+
+    protected final AtomicReference<State> currentState = new AtomicReference<>(State.CLOSED);
+    protected volatile long nextRetryTimestamp;
+
+    public AbstractCircuitBreaker(DegradeRule rule) {
+        this(rule, EventObserverRegistry.getInstance());
+    }
+
+    AbstractCircuitBreaker(DegradeRule rule, EventObserverRegistry observerRegistry) {
+        AssertUtil.notNull(observerRegistry, "observerRegistry cannot be null");
+        if (!DegradeRuleManager.isValidRule(rule)) {
+            throw new IllegalArgumentException("Invalid DegradeRule: " + rule);
+        }
+        this.observerRegistry = observerRegistry;
+        this.rule = rule;
+        this.retryTimeoutMs = rule.getTimeWindow() * 1000;
+    }
+
+    @Override
+    public DegradeRule getRule() {
+        return rule;
+    }
+
+    @Override
+    public State currentState() {
+        return currentState.get();
+    }
+
+    @Override
+    public boolean tryPass() {
+        // Template implementation.
+        if (currentState.get() == State.CLOSED) {
+            return true;
+        }
+        if (currentState.get() == State.OPEN) {
+            // For half-open state we allow a request for trial.
+            return retryTimeoutArrived() && fromOpenToHalfOpen();
+        }
+        return false;
+    }
```

Comment:
If thare are two breakers and curState is open, one breaker tryPass and change state to half open, but other breaker tryPass return false, it is right that no request allow. but in below requqests will no allow pass because two breakers  can't both be true.

Reply:
@sczyh30  Any idea for the scene？Was I wrong？

Reply:
I think it's ok for this scenario.

B1(open) & B2(open)   -> no request at all
B1(open -> half) & B2(open)   -> no request at all, but B1 may transform to closed leaving B2 alone, that is:
B1(closed) & B2(open)

Reply:
Sorry, i cannot find how to change half-open to close.
Method `fromHalfOpenToClose` is invoked in `DegradeSlot#exist` when no BlockingException. But if B1(open -> half) & B2(open), the follow requests check by `CircuitBreaker#tryPass` must return false then throw DegradeException.

Reply:
Requests is blocked by `tryPass()` in abstract parent and the state of breakers are maintained in different implementations when requests complete (post-request).

So they are not collided.

Reply:
可能我说的不清楚。
假如此时 B1(open) & B2(open)，B1,B2的 recoveryTimeoutMs 不一样，下一个请求导致B1(half)，而B2由于还未达到 timeout 时间，仍保持 B2(open)，故 tryPass 不通过进而抛出 BlockException. 在 DegradeSlot#exit 中仅对非 BlockException 的情况维护 Breaker 状态，故此时 B1(half) & B2(open).

```java
public void exit(Context context, ResourceWrapper r, int count, Object... args) {
        Entry curEntry = context.getCurEntry();
        if (curEntry.getBlockError() != null) {
            fireExit(context, r, count, args);
            return;
        }
        ...
        if (curEntry.getBlockError() == null) {
          ...
          for (CircuitBreaker circuitBreaker : circuitBreakers) {
                circuitBreaker.onRequestComplete(rt, error);
            }
        }
       ...
}
```java

后面的请求，由于 B1 不符合 [open, close] 状态，tryPass 必然为 false 又抛 BlockException，进入循环。

> B1(open -> half) & B2(open) -> no request at all, but B1 may transform to closed leaving B2 alone, that is:
B1(closed) & B2(open)

这个我没找到在什么哪里有个逻辑实现。

Reply:
噢我懂你关注的点了，不过这个场景中的处理有问题吗？
B1/B2同时对资源做处理（例如一个针对RT一个针对ErrorRate），这时一个关断一个从关断进入半恢复状态，那么单个的状态是不是有及时恢复，理论上没这么重要，因为其中之一关断则资源关断，两者较长恢复时间者进入半开则两者恢复通路，对于不同管控指标同一资源的断路器这里貌似不存在问题。是这样吗？

Reply:
不对吧。
> 后面的请求，由于 B1 不符合 [open, close] 状态，tryPass 必然为 false 又抛 BlockException，进入循环

就算 B2 可以进入半开，但是 B1(half) 也会因为上面的判断导致抛出 BlockException，后续在 exit 方法中也不会尝试更改断路器的状态了

Reply:
在tryPass中，

open -> false
closed -> true
half -> retryTimeout && halfToOpen

这里并不是说非open非close一定false的，是我理解错了吗？

Reply:
```java
public boolean tryPass() {
        // Template implementation.
        if (currentState.get() == State.CLOSED) {
            return true;
        }
        if (currentState.get() == State.OPEN) {
            // For half-open state we allow a request for trial.
            return retryTimeoutArrived() && fromOpenToHalfOpen();
        }
        return false;
    }
```
我已经自我怀疑了🤨😢  是我理解错了吗？？

Reply:
> ```java
> ```java
> public boolean tryPass() {
>         // Template implementation.
>         if (currentState.get() == State.CLOSED) {
>             return true;
>         }
>         if (currentState.get() == State.OPEN) {
>             // For half-open state we allow a request for trial.
>             return retryTimeoutArrived() && fromOpenToHalfOpen();
>         }
>         return false;
>     }
> ```
> 
> 
> 我已经自我怀疑了🤨😢  是我理解错了吗？？
> ```

emmm.........仔细想了下，你考虑的问题的确存在，而且貌似还存在其它限流规则的问题，因为这里并没有判断是不是DegradeException，待我下午拿Test测一下

Reply:
@sczyh30 Hi, Eric

Here're some boundaries which leads to new issues while we should discuss more how to make the recovery logic better:

1. Multiple degrading rules for same resource(Especially after introducing prefix/postfix matching)
2. Another rule after degrading rule in half-open state(Though there's no such kind of rule but it's possible in future)

These will lead to the incorrect workflow:

entry:   R0(open -> half) -> R1(block)
exit: XXXBlockException -> skip the detection for stalling half state

For straightforward patch i think the `onComplete` method is intentioned to be call after a successfully transforming into half. We should make the exiting logic more simple and clear.

Reply:
@jasonjoo2010 @wavesZh Yes, it's a fatal bug indeed and should be resolved immediately. However we need a better temporary workaround for this (for the half-open case). I'll merge this PR first and we may discuss it in a new issue. It's a common problem for all rules that have their own metrics and rely on `onComplete`.

Reply:
See #1638